enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,1167 @@
+service exists. Services that exist before the user-defined network is assigned to a namespace will result in
+OVN-Kubernetes executing a re-sync on all services in that namespace, and updating all load balancers. Keep in mind that
+pods must not exist in the namespace when the namespace is assigned to a new network or the new network assignment will
+not be accepted by OVN-Kubernetes.
+
+Services in a user-defined network will be reachable by other namespaces that share the same network.
+
+As previously mentioned, Kubernetes API and DNS services will be accessible by all pods.
+
+Endpoint slices will provide the IPs of the cluster default network in Kubernetes API. For this implementation the required
+endpoints are those IP addresses which reside on the user-defined primary network. In order to solve this problem,
+OVN-Kubernetes may create its own endpoint slices or may choose to do dynamic lookups at runtime to map endpoints to
+their primary IP address. Leveraging a second set of endpoint slices will be the preferred method, as it creates less
+indirection and gives explicit Kube API access to what IP addresses are being used by OVN-Kubernetes.
+
+Kubelet health checks to pods are queried via the cluster default network. When endpoints are considered unhealthy they
+will be removed from the endpoint slice, and thus their primary IP will be removed from the OVN load balancer. However,
+it is important to note that the healthcheck is being performed via the cluster default network interface on the pod, 
+which ensures the application is alive, but does not confirm network connectivity of the primary interface. Therefore,
+there could be a situation where OVN networking on the primary interface is broken, but the default interface continues
+to work and reports 200 OK to Kubelet, thus rendering the pod serving in the endpoint slice, but unable to function.
+Although this is an unlikely scenario, it is good to document.
+
+### Network Policy
+
+Network Policy will be fully supported for user-defined primary networks as it is today with the cluster default network.
+However, configuring network policies that allow traffic between namespaces that connect to different user-defined
+primary networks will have no effect. This traffic will not be allowed, as the networks have no connectivity to each other.
+These types of policies will not be invalidated by OVN-Kubernetes, but the configuration will have no effect. Namespaces
+that share the same user-defined primary network will still benefit from network policy that applies access control over
+a shared network. Additionally, policies that block/allow cluster egress or ingress traffic will still be enforced for
+any user-defined primary network.
+
+### API Extensions
+
+The main API extension here will be a namespace scoped network CRD as well a cluster scoped network CRD. 
+These CRDs will be registered by Cluster Network Operator (CNO). 
+
+#### CRDs
```

Comment:
Is it fair to say that these new CRDs are meant to be exposed directly to OpenShift users, as the main way to define their primary networks? Or are we planning to wrap it in another CRD?

If the former, I think it's critical to have the OpenShift API team review and accept the CRD API. Otherwise we risk that will will not be allowed to use these new APIs as a core part of OpenShift.

I'm saying this because we are now facing similar issue with  kubernetes-nmstate API.

Reply:
directly exposed to OpenShift users, but they exist as part of the OVN-Kubernetes API and not OpenShift API. Therefore, there wont be any OpenShift API change for this. The CRDs for downstream are stored in the CNO repo and registered by CNO. That being said, we still should have the API team give us their feedback. We plan to do this after internally we are in agreement on what we think the API should look like, then can ask for their feedback.

Reply:
Understood, thanks